src/ring_theory/subsemiring.lean

```diff
@@ -369,6 +420,45 @@ lemma closure_eq_of_le {s : set R} {t : subsemiring R} (h₁ : s ⊆ t) (h₂ :
   closure s = t :=
 le_antisymm (closure_le.2 h₁) h₂
 
+end subsemiring
+
+namespace submonoid
+
+lemma to_subsemiring_coe :
+  (M.to_subsemiring : set R) = add_submonoid.closure (M : set R) := rfl
+
+lemma to_subsemiring_to_add_submonoid :
+  M.to_subsemiring.to_add_submonoid = add_submonoid.closure (M : set R) := rfl
+
+lemma to_subsemiring_eq_closure : M.to_subsemiring = subsemiring.closure (M : set R) :=
+begin
+  ext,
+  refine ⟨λ hx, _, λ hx, (subsemiring.mem_closure.mp hx) M.to_subsemiring (λ s sM, _)⟩;
+  rintros - ⟨H1, rfl⟩;
+  rintros - ⟨H2, rfl⟩,
+  { exact add_submonoid.mem_closure.mp hx H1.to_add_submonoid H2 },
+  { exact H2 sM }
+end
+
+end submonoid
+
+namespace subsemiring
+
+/-- The `subsemiring.closure` of a multiplicative submonoid coincides with the `subsemiring`
+generated by the submonoid itself. -/
+lemma to_subsemiring_to_add_submonoid :
+  M.to_subsemiring.to_add_submonoid = add_submonoid.closure (M : set R) := rfl
+
+/-- The elements of the subsemiring closure of `M` are exactly the elements of the additive closure
+of a multiplicative submonoid `M`. -/
+lemma mem_closure_iff {s : set R} {x} :
+  x ∈ closure s ↔ x ∈ add_submonoid.closure (submonoid.closure s : set R) :=
+begin
+  rw [← add_submonoid.mem_coe, ← submonoid.to_subsemiring_coe, submonoid.to_subsemiring_eq_closure],
+  refine ⟨λ hx, closure_mono (submonoid.subset_closure) hx, λ hx, (mem_closure.mp hx) _ _⟩,
+  exact λ y hy, (submonoid.mem_closure.mp hy) (subsemiring.to_submonoid (closure s)) subset_closure
```

Comment:
I think there's a lemma hiding in here worth stating separately:
```lean
@[simp]
lemma subsemiring.closure_submonoid_closure : closure ↑(submonoid.closure s) = closure s`
```

Reply:
I liked and added this lemma!